spec/datetimeformat.html

```diff
@@ -80,7 +80,7 @@ <h1>InitializeDateTimeFormat ( _dateTimeFormat_, _locales_, _options_ )</h1>
         1. If _hour12_ is *true*, then
           1. If _hcDefault_ is *"h11"* or *"h23"*, let _hc_ be *"h11"*. Otherwise, let _hc_ be *"h12"*.
         1. Else if _hour12_ is *false*, then
-          1. If _hcDefault_ is *"h11"* or *"h23"*, let _hc_ be *"h23"*. Otherwise, let _hc_ be *"h24"*.
+          1. Let _hc_ be *"h23"*.
```

Comment:
Would better logic be,

```
1. If _hour12_ is *true*, then
  1. If _hcDefault_ is *"h23"* or *"h24"*, let _hc_ be *"h12"*. Otherwise, let _hc_ be _hcDefault_.
1. Else if _hour12_ is *false*, then
  1. If _hcDefault_ is *"h11"* or *"h12"*, let _hc_ be *"h23"*. Otherwise, let _hc_ be _hcDefault_.
```

Reply:
this will not work for
```js
(new Intl.DateTimeFormat("ja", {hour: "numeric", hour12: true})).resolvedOptions().hourCycle;
```
_hcDefault_ for ja is h23 but if hour12:true it should be h11, not h23 nor h12
your logic will make it h23.
also there are more problem for the case if _hour12_ is *false*

Reply:
Can you cite the CLDR data that shows that "ja" in 12-hour mode is expected to use h11, and likewise any locale in 24-hour mode that is expected to use h24?

Reply:
> Can you cite the CLDR data that shows that "ja" in 12-hour mode is expected to use h11, and likewise any locale in 24-hour mode that is expected to use h24?

That is not what my PR proposed.

What my PR proposed is 
the following mapping

| hour12  | undefined | false  |   true |  
|---|---|---|---| 
| h11  |  h11  | h23 |  h11  | 
| h12 | h12  |  h23* |  h12  |
| h23  | h23  |  h23 |  h11 |
| h24  |  h24  |  h24 |  h12  |

The differences from the status quo is while the hour cycle is h12 and hour12:false  instead of making it h24 (which show 24:30:59) we show (00:30:59) and that impact mostly "en-US" or "en-CA" , "fr-CA" locales

Reply:
about japanese, I am not sure this count as what you asked
https://github.com/unicode-org/cldr/blob/main/common/main/ja.xml#L3059
```xml
						<dateFormatItem id="h">aK時</dateFormatItem>
						<dateFormatItem id="H">H時</dateFormatItem>
```
h-> aK =>  use h11 when asking for not in 24 hour mode
H->H => use h23 when asking for in 24 hour mode

Reply:
Thanks for the Japanese citation

I'm not super happy with the spec codifying this table. The only restriction should be that you get either h11 or h12 in 12-hour mode or either h23 or h24 in 24-hour mode. Which one should be ILD.

Reply:
I understand why you are not happen with the spec codifying this table. 

" The only restriction should be that you get either h11 or h12 in 12-hour mode or either h23 or h24 in 24-hour mode"
YES, but that is not what the current spec text specify.  The current spec text specifiy a logic  to resolve that instead purely depend on the  CLDR to resolve that.

Reply:
Alternatively we could add two new fields to [LocaleData](https://tc39.es/ecma402/#sec-intl.datetimeformat-internal-slots) to describe the preferred hour-cycles for 12- and 24-hour clocks.

> - [[LocaleData]].[[\<locale\>]].[[hourCycle12]] must be a String value equal to "h11" or "h12". 
> - [[LocaleData]].[[\<locale\>]].[[hourCycle24]] must be a String value equal to "h23" or "h24". 


And then rewrite steps 24-26 as:
> 25. If hour12 is true, then
>     a. Let hc be dataLocaleData.[[hourCycle12]].
> 26. Else if hour12 is false, then
>     a. Let hc be dataLocaleData.[[hourCycle24]].
> 27. Else,
>     a. Assert: hour12 is undefined.
>     b. Let hc be r.[[hc]].
>     c. If hc is null, set hc to dataLocaleData.[[hourCycle]].

Reply:
> Alternatively we could add two new fields to [LocaleData](https://tc39.es/ecma402/#sec-intl.datetimeformat-internal-slots) to describe the preferred hour-cycles for 12- and 24-hour clocks.
> 
> >

hum.... that will make it more a data driven approach ( which is good). Shane- what do you think?

Reply:
+1 on @anba's approach.